### PR TITLE
[Merged by Bors] - feat: the Unfolding Trick

### PR DESCRIPTION
**Pull request by AlexKontorovich**
This PR is a forward port of mathlib3 PR !3#18863

Co-authored-by: Heather Macbeth <25316162+hrmacbeth@users.noreply.github.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:282]:
```suggestion
    [NormedSpace ℝ K] [μ'.IsAddRightInvariant] {f : G' → K}
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:236]:
```suggestion
    [NormedSpace ℝ K] [μ.IsMulRightInvariant] {f : G → K}
```

File: Mathlib/GroupTheory/GroupAction/Group.lean
Comment (ADedecker) [inline Mathlib/GroupTheory/GroupAction/Group.lean:271]:
```suggestion
@[simps!]
def DistribMulAction.toAddEquiv₀ {α : Type _} (β : Type _) [GroupWithZero α] [AddMonoid β]
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1886]: This is exactly [MeasureTheory.AEStronglyMeasurable.mono'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.html#MeasureTheory.AEStronglyMeasurable.mono') right?

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ := by
  rw [← (Equiv.inv G).tsum_eq, lintegral_eq_tsum' h f]; rfl
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:435]:
```suggestion
    (f : α → E) (hf : Integrable f μ) : ∫ x, f x ∂μ = ∑' g : G, ∫ x in s, f (g • x) ∂μ := by
  rw [← (Equiv.inv G).tsum_eq, integral_eq_tsum' h f hf]; rfl
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:124]:
```suggestion
    [SecondCountableTopology (G ⧸ Γ)] (K : PositiveCompacts (G ⧸ Γ)) [Subgroup.Normal Γ]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (ADedecker) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:150]:
```suggestion
    (K : PositiveCompacts (G ⧸ Γ)) [Subgroup.Normal Γ] [μ.IsHaarMeasure]
```

**Comment by ADedecker** [comment]
Oh sorry, I had somehow missed the fact that this is a forward-port :facepalm: I'll delete some of my comments, the one left should still be applicable.

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:252]:
```suggestion
    · refine (AEStronglyMeasurable.mono' hg ?_).comp_measurable meas_π
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean:1891]:
```suggestion
```

File: Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Function/StronglyMeasurable/Basic.lean]: Yes, thanks! Deleted and updated.

**Comment by AlexKontorovich** [comment]
> 

Oh I see; I thought that was just for new files being moved, not files that were already ported but are now being modified. I tried clicking the link, but didn't follow which of that applied to this port? Which headers need to change, all of them?

**Comment by ADedecker** [comment]
Those of all files modified by your PR, yes. To find the right header, you should look on the [out-of-sync page](https://leanprover-community.github.io/mathlib-port-status/out-of-sync) and click on the file you are porting. Then the left column shows the changes in mathlib3 since the last synchronization. When you hover over he change you are porting, it highlights the corresponding change in the middle column (which is the output of mathport), and looking at the corresponding diff tells you what to change.

**Comment by AlexKontorovich** [comment]
> Those of all files modified by your PR, yes. To find the right header, you should look on the [out-of-sync page](https://leanprover-community.github.io/mathlib-port-status/out-of-sync) and click on the file you are porting. Then the left column shows the changes in mathlib3 since the last synchronization. When you hover over he change you are porting, it highlights the corresponding change in the middle column (which is the output of mathport), and looking at the corresponding diff tells you what to change.

Done, thanks!

**Review by YaelDillies — COMMENTED** [review]
This is a faithful forward-port, but where are all the `#align`s gone?

Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do not merge this**.

maintainer merge

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (ericrbg) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]: what happens with `@[to_additive]` automatically here? should this be a bug report?

Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]: I think there are just too many things going on at once for `@[to_additive]` to work out which things should be additivized and which should remain the same...

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (sgouezel) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: indentation is wrong here

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1557]: indentation is wrong here.

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (sgouezel) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1298]: can you replace all `Type _` in the PR with `Type*`?

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: sorry can you please tell me what's wrong? I don't see. Thanks!

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [reply Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1557]: Here too - what should the right indentation be?

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1552]:
```suggestion
variable {G : Type*} [AddGroup G] {Γ : AddSubgroup G}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1532]:
```suggestion
variable {G : Type*} [Group G] {Γ : Subgroup G}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1469]:
```suggestion
variable {M : Type*} [TopologicalSpace M] [AddCommMonoid M] [T2Space M] {R : Type*}
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1314]:
```suggestion
lemma tsum_const_smul'' {γ : Type*} [DivisionRing γ] [Module γ α] [ContinuousConstSMul γ α]
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1298]:
```suggestion
lemma tsum_const_smul' {γ : Type*} [Group γ] [DistribMulAction γ α] [ContinuousConstSMul γ α]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:280]:
```suggestion
lemma QuotientAddGroup.integral_mul_eq_integral_automorphize_mul {K : Type*} [NormedField K]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:267]:
```suggestion
variable {G' : Type*} [AddGroup G'] [MeasurableSpace G'] [TopologicalSpace G']
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:235]:
```suggestion
lemma QuotientGroup.integral_mul_eq_integral_automorphize_mul {K : Type*} [NormedField K]
```

File: Mathlib/MeasureTheory/Measure/Haar/Quotient.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Measure/Haar/Quotient.lean:216]:
```suggestion
lemma QuotientGroup.integral_eq_integral_automorphize {E : Type*} [NormedAddCommGroup E]
```

File: Mathlib/GroupTheory/GroupAction/Group.lean
Comment (AlexKontorovich) [inline Mathlib/GroupTheory/GroupAction/Group.lean:271]:
```suggestion
def DistribMulAction.toAddEquiv₀ {α : Type*} (β : Type*) [GroupWithZero α] [AddMonoid β]
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (AlexKontorovich) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: Ok I think I've got it.

Comment (jcommelin) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: The indentation level of the type (aka statement) has changed with the switch to Lean 4.
The convention is now to indent by 4 spaces on a new line after the `:`.

Comment (jcommelin) [reply Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]: If the statement is short, you can of course still write it on the same line as the `:` with only 1 space in between.

Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:258]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ := 
  (lintegral_eq_tsum' h f).trans ((Equiv.inv G).tsum_eq (fun g ↦ ∫⁻ (x : α) in s, f (g • x) ∂μ))
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1553]:
```suggestion
    QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
      = g • (QuotientAddGroup.automorphize f : G ⧸ Γ → M) :=
  AddAction.automorphize_smul_left f g
```

File: Mathlib/Topology/Algebra/InfiniteSum/Basic.lean
Comment (AlexKontorovich) [inline Mathlib/Topology/Algebra/InfiniteSum/Basic.lean:1555]:
```suggestion
    QuotientAddGroup.automorphize ((g ∘ (@Quotient.mk' _ (_))) • f)
      = g • (QuotientAddGroup.automorphize f : G ⧸ Γ → M) :=
  AddAction.automorphize_smul_left f g
```

File: Mathlib/MeasureTheory/Group/FundamentalDomain.lean
Comment (AlexKontorovich) [inline Mathlib/MeasureTheory/Group/FundamentalDomain.lean:257]:
```suggestion
    ∫⁻ x, f x ∂μ = ∑' g : G, ∫⁻ x in s, f (g • x) ∂μ :=
```

**Comment by eric-wieser** [comment]
> Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do note merge this**.

We should be keeping the aligns, because otherwise there wasn't much point merging the mathlib3 version.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by eric-wieser** [comment]
@jcommelin, I take it that's a statement that `#align`s don't matter any more?

**Comment by eric-wieser** [comment]
bors r-

(I think Johan's merge was in response to a maintainer merge comment that Yael deleted?)

**Comment by Ruben-VandeVelde** [comment]
> > Whichever maintainer is reading this? Should we keep on adding the `#align`s? If so, please **do note merge this**.
> 
> We should be keeping the aligns, because otherwise there wasn't much point merging the mathlib3 version.

I thought the main point of merging things into mathlib 3 at this point was to take advantage of mathport?

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**
